### PR TITLE
Enable grain input pipeline save and restore for distillation.

### PR DESCRIPTION
**Pull request by gagika**
# Description

This PR enables input pipeline checkpointing for the Distillation workflow, specifically supporting Grain datasets. It ensures that when a distillation run is interrupted and resumed, the data iterator restores its state correctly instead of restarting from the beginning.

**Key Changes:**
* **Input Pipeline Checkpointing:** Implemented `MaxTextCheckpointManager` in `distillation_utils.py`. This extends the Tunix checkpoint manager to utilize MaxText's native `GrainCheckpointHandler`, allowing the iterator state to be saved and restored alongside model parameters.
* **Refactoring:** Extracted distillation-specific utilities (data adapters, custom input structs, logit strategies) from `train_distill.py` into a new module `distillation_utils.py` to improve code organization and modularity.
* **Integration:** Updated `train_distill.py` to initialize the custom checkpoint manager and automatically restore the iterator state if a checkpoint exists.

# Tests

**Unit Tests:**
* Added `tests/unit/distillation_checkpointing_test.py` to verify that `MaxTextCheckpointManager` correctly saves and restores iterator state using `GrainCheckpointHandler`.
* Ran existing tests: `tests/unit/train_distill_test.py`.

**Integration Tests:**
Verified manually by running the distillation trainer with and without the Grain pipeline enabled. Confirmed that the training loop initializes and runs.

Command used:
```bash
python3 src/maxtext/trainers/post_train/distillation/train_distill.py \
    src/maxtext/configs/post_train/distillation.yml \
    run_name=$RUN_NAME \
    base_output_directory=${BASE_OUTPUT_DIRECTORY} \
    hf_access_token=${HF_TOKEN}
```   
 

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Review by aireenmei — APPROVED** [review]
Thanks for the work!
